src/deps.ts

```diff
@@ -1,36 +1,13 @@
-const kModuleError = Symbol('moduleError');
+import { MongoError } from './error';
 
-function makeErrorModule(error: any) {
-  const props = error ? { [kModuleError]: error } : {};
-  return new Proxy(props, {
-    get: (_: any, key: any) => {
-      if (key === kModuleError) {
-        return error;
-      }
-      throw error;
-    },
-    set: () => {
-      throw error;
-    }
-  });
+interface OptionalModule {
+  kModuleError: MongoError;
 }
 
-let Kerberos = makeErrorModule(
-  new Error(
-    'Optional module `kerberos` not found. Please install it to enable kerberos authentication'
-  )
-);
-
-try {
-  Kerberos = require('kerberos');
-} catch (_) {} // eslint-disable-line
-
-let Snappy = makeErrorModule(
-  new Error('Optional module `snappy` not found. Please install it to enable snappy compression')
-);
-
-try {
-  Snappy = require('snappy');
-} catch (_) {} // eslint-disable-line
-
-export { Kerberos, Snappy, kModuleError };
+export function optionalRequire<T>(moduleName: string): T | OptionalModule {
+  try {
+    return require(moduleName);
+  } catch {
+    return { kModuleError: new MongoError(`Optional Module ${moduleName} not found`) };
+  }
+}
```

Comment:
This is my typescript take on the optional dependency, you can jump to `gssapi.ts` to see how it looks when called. I have to use a string for the prop that holds the error that might be there since since

```ts
if ('kModuleError' in Kerberos) {
  // Kerberos is `OptionalModule` type
}
// Kerberos is `typeof import('kerberos')` type
```
the `in` type guard requires a string literal on the left hand side. I think this will work for us, we were using a Symbol for kModuleError for maximum avoiding of collision, I think with TS in play if any of the modules got updated to have a property `'kModuleError'` then it can't have the type MongoError so we would get conflicts and be able to dodge bugs there.

Reply:
I think the approach might work, but I'm concerned about what happens at the call site when you do something like `typeof import('kerberos')` and `kerberos` is not installed. Can we maybe leave these changes out of this PR and try to address them by themselves in a follow-on PR?